go/vt/dbconnpool/connection_pool.go

```diff
@@ -65,15 +65,15 @@ func NewConnectionPool(name string, capacity int, idleTimeout time.Duration) *Co
 		return cp
 	}
 	usedNames[name] = true
-	stats.Publish(name+"Capacity", stats.IntFunc(cp.Capacity))
-	stats.Publish(name+"Available", stats.IntFunc(cp.Available))
-	stats.Publish(name+"Active", stats.IntFunc(cp.Active))
-	stats.Publish(name+"InUse", stats.IntFunc(cp.InUse))
-	stats.Publish(name+"MaxCap", stats.IntFunc(cp.MaxCap))
-	stats.Publish(name+"WaitCount", stats.IntFunc(cp.WaitCount))
-	stats.Publish(name+"WaitTime", stats.DurationFunc(cp.WaitTime))
-	stats.Publish(name+"IdleTimeout", stats.DurationFunc(cp.IdleTimeout))
-	stats.Publish(name+"IdleClosed", stats.IntFunc(cp.IdleClosed))
+	stats.NewGaugeFunc(name+"Capacity", "Connection pool capacity", cp.Capacity)
+	stats.NewGaugeFunc(name+"Available", "Connection pool available", cp.Available)
+	stats.NewGaugeFunc(name+"Active", "Connection pool active", cp.Active)
+	stats.NewGaugeFunc(name+"InUse", "Connection pool in-use", cp.InUse)
+	stats.NewGaugeFunc(name+"MaxCap", "Connection pool max cap", cp.MaxCap)
+	stats.NewGaugeFunc(name+"WaitCount", "Connection pool wait count", cp.WaitCount)
+	stats.NewDurationFunc(name+"WaitTime", "Connection pool wait time", cp.WaitTime)
+	stats.NewDurationFunc(name+"IdleTimeout", "Connection pool idle timeout", cp.IdleTimeout)
```

Comment:
Ugh... looking at this deeper I'm afraid to say that we need the ability for a duration to map to either a prom counter or a prom gauge.

In these two examples, WaitTime is a counter since (like WaitCount) it is only ever incremented, while IdleTimeout is a gauge since it represents the configured timeout.

To that end we have a couple options... the simplest would be to have `NewDurationGaugeFunc` and `NewDurationCounterFunc` which bloats the interface but at the same time is most consistent with the other cases.

Another option would be to overload the existing `stats.NewGaugeFunc` and `stats.NewCounterFunc` interfaces with an optional conversion function which would implement the mapping, much like how the old API was.

This way the caller API would be:
`stats.NewCounterFunc(name+"WaitTime", "Connection pool wait time", stats.DurationFunc(cp.WaitTime))`

Then just like before, the implementation of `stats.DurationFunc` would allow either expvar to get the String version or prom to get the underlying Duration.

FWIW I _think_ that would allow the prom implementation to call `duration.Seconds` instead of going to nanoseconds and dividing which would reduce rounding errors.

Reply:
re: WaitCount 

ohhhhh! I didn't peel the next layer of the onion. Wow! 

``` 
func (rp *ResourcePool) recordWait(start time.Time) {
        rp.waitCount.Add(1)
        rp.waitTime.Add(time.Now().Sub(start))
}
```

It only ever increments! Also, this seems like WaitCount & WaitTime should be condensed into a histogram/time.

Reply:
But I think for now keeping this separated and making DurationFunc support both counters and gauges sounds good.